source/common/upstream/eds.cc

```diff
@@ -1,53 +1,54 @@
-#include "common/upstream/sds.h"
+#include "common/upstream/eds.h"
 
-#include <cstdint>
-#include <string>
-#include <vector>
-
-#include "common/http/headers.h"
-#include "common/json/config_schemas.h"
 #include "common/network/address_impl.h"
 #include "common/network/utility.h"
+#include "common/upstream/sds_subscription.h"
 
 namespace Envoy {
 namespace Upstream {
 
-SdsClusterImpl::SdsClusterImpl(const Json::Object& config, Runtime::Loader& runtime,
+EdsClusterImpl::EdsClusterImpl(const Json::Object& config, Runtime::Loader& runtime,
                                Stats::Store& stats, Ssl::ContextManager& ssl_context_manager,
                                const SdsConfig& sds_config, const LocalInfo::LocalInfo& local_info,
                                ClusterManager& cm, Event::Dispatcher& dispatcher,
                                Runtime::RandomGenerator& random)
-    : BaseDynamicClusterImpl(config, runtime, stats, ssl_context_manager),
-      RestApiFetcher(cm, sds_config.sds_cluster_name_, dispatcher, random,
-                     sds_config.refresh_delay_),
-      local_info_(local_info), service_name_(config.getString("service_name")) {}
-
-void SdsClusterImpl::parseResponse(const Http::Message& response) {
-  Json::ObjectSharedPtr json = Json::Factory::loadFromString(response.bodyAsString());
-  json->validateSchema(Json::Schema::SDS_SCHEMA);
+    : BaseDynamicClusterImpl(config, runtime, stats, ssl_context_manager), local_info_(local_info),
+      cluster_name_(config.getString("service_name")) {
+  // TODO(htuch): This is where the v2 Subscription implementations will be constructed based on
+  // config. Today, we just reuse the v1 SDS config and an adapter Subscription to the v1 REST JSON
+  // API.
+  subscription_.reset(new SdsSubscription(info_->stats(), sds_config, cm, dispatcher, random));
+}
+
+void EdsClusterImpl::initialize() { subscription_->start({cluster_name_}, *this); }
+
+bool EdsClusterImpl::onConfigUpdate(const ResourceVector& resources) {
```

Comment:
What I meant in the 2nd comment is that whether you use bool return or exceptions, you need to handle the initialize paths in error cases: https://github.com/htuch/envoy/blob/569e3cc72beca28485d6e64491601ac3d060d5ff/source/common/upstream/eds.cc#L93

Basically, just need to make sure that we still initialize if something fails so we don't block server startup.

And yes, I think exception based error handling will make the code much simpler and easier to understand ultimately. (This is almost always going to be my answer, so probably time to lose the exception allergy :) )